_posts/2018-02-27-django-k8s-elb-health-checks.md

```diff
@@ -0,0 +1,85 @@
+---
+layout: post
+title: Django, K8s, and ELB Health checks
+author: Paul McLanahan
+excerpt_separator: <!--more-->
+---
+
+As you may have seen in several of our
+[SRE](https://mozilla.github.io/meao/2018/02/16/sre-status/)
+[status](https://mozilla.github.io/meao/2018/01/23/sre-status/)
+[reports](https://mozilla.github.io/meao/2017/12/05/sre-status/)
+we're moving all of our webapp hosting from [Deis][] to [Kubernetes][] (k8s). As part of that
+we've also been doing some additional thinking about the security of our deployments.
+One thing we've not done as good a job as we should is with [Django's `ALLOWED_HOSTS` setting][].
+We should have been adding all possible hosts to that list, but it seems we used to occassionally
+leave it set to `['*']`. This isn't great, but also isn't the end-of-the-world since
+we don't knowingly construct URLs using the info sent via the `Host` header. In an effort
+to cover all bases we've decided to improve this. Unfortunately our particular combination
+of technologies doesn't make this as
+[easy as we thought it would](https://github.com/mozmeao/infra/issues/735) (story of our lives).
+
+<!--more-->
+
+## AWS ELB Health Checks
+
+Here's the thing: Amazon Web Services' (AWS) Elastic Load Balancers (ELB) do not have many options for
+their health checks. These checks ensure that your app on a particular node in your cluster
+is working as expected. If the check fails the ELB will remove the node from the list of nodes
+to which it will route requests for your app. However, because it's hitting the nodes directly
+it doesn't rely on DNS and directly requests the IP address and port, and it doesn't allow you
+to specify custom headers (e.g. the `Host` header). It also can't do HTTPS because we terminate
+TLS connections at the ELB, so the app nodes speak only plain HTTP back to the ELB. All of that
+means that our health check endpoint needs to do two unique things: allow HTTP connection and
+allow the IP address that the ELB requests as a valid `Host` header. The first bit is easy enough
+when using Django's in-built `SecurityMiddleware` since it supports the `SECURE_REDIRECT_EXEMPT`
+setting. It's this second requirement that gets more interesting when combined with k8s.
+
+## K8s Routing
+
+The way I understand it (and I'm admittedly no expert) is that k8s (at least the way we use it)
+sets up a "node port" per app (or namespace). To hit that app you can hit any node in the cluster
```

Comment:
`NodePort` is how it's referred to in the [K8s docs](https://kubernetes.io/docs/concepts/services-networking/service/)